File: src/main/java/com/company/erp/features/users/dto/UpdateRoleRequest.java
package com.company.erp.features.users.dto;

import jakarta.validation.constraints.Size;

import java.util.Set;

public record UpdateRoleRequest(
        @Size(max = 128) String name,
        @Size(max = 255) String description,
        /** Replaces the role's permission set. {@code null} means "leave permissions untouched." */
        Set<String> permissions
) {
}

File: app/Features/Users/Requests/UpdateRoleRequest.php
<?php

namespace App\Features\Users\Requests;

use App\Support\Http\ApiFormRequest;

class UpdateRoleRequest extends ApiFormRequest
{
    public function rules(): array
    {
        return [
            'name' => ['sometimes', 'nullable', 'string', 'max:128'],
            'description' => ['sometimes', 'nullable', 'string', 'max:255'],
            'permissions' => ['sometimes', 'nullable', 'array'],
            'permissions.*' => ['string'],
        ];
    }
}
